File: reference/session_replay.py
#!/usr/bin/env python3
"""Portable, bounded MotionBricks observational replay format."""

from __future__ import annotations

import struct
from dataclasses import dataclass
from pathlib import Path

import numpy as np


MAGIC = b"MBRPLY1\0"
VERSION = 1
HEADER = struct.Struct("<8s8I")
TARGET_FRAMES = 4


@dataclass
class Replay:
    fps: int
    parents: np.ndarray
    modes: np.ndarray
    frame_plans: np.ndarray
    qpos: np.ndarray
    joint_positions: np.ndarray
    plan_frames: np.ndarray
    plan_modes: np.ndarray
    plan_valid_lengths: np.ndarray
    target_positions: np.ndarray

    @property
    def frame_count(self) -> int:
        return int(self.qpos.shape[0])

    @property
    def joint_count(self) -> int:
        return int(self.parents.shape[0])

    @property
    def qpos_count(self) -> int:
        return int(self.qpos.shape[1])

    @property
    def plan_count(self) -> int:
        return int(self.plan_frames.shape[0])


def _array(value, dtype, shape: tuple[int, ...], name: str) -> np.ndarray:
    result = np.asarray(value, dtype=dtype)
    if result.shape != shape:
        raise ValueError(f"{name} has shape {result.shape}, expected {shape}")
    if np.issubdtype(result.dtype, np.floating) and not np.isfinite(result).all():
        raise ValueError(f"{name} contains a non-finite value")
    return np.ascontiguousarray(result)


def validate(replay: Replay) -> Replay:
    frames, joints, qpos, plans = (
        replay.frame_count, replay.joint_count, replay.qpos_count, replay.plan_count
    )
    if replay.fps <= 0 or frames <= 0 or joints <= 0 or joints > 64 or qpos <= 0 or plans <= 0:
        raise ValueError("replay dimensions are outside supported bounds")
    replay.parents = _array(replay.parents, "<i4", (joints,), "parents")
    if replay.parents[0] != -1:
        raise ValueError("root parent must be -1")
    for joint, parent in enumerate(replay.parents):
        if joint > 0 and (parent < 0 or parent >= joint):
            raise ValueError(f"invalid parent {parent} for joint {joint}")
    replay.modes = _array(replay.modes, "<i4", (frames,), "modes")
    replay.frame_plans = _array(replay.frame_plans, "<i4", (frames,), "frame_plans")
    if np.any(replay.frame_plans < -1) or np.any(replay.frame_plans >= plans):
        raise ValueError("frame plan index is out of bounds")
    replay.qpos = _array(replay.qpos, "<f4", (frames, qpos), "qpos")
    replay.joint_positions = _array(
        replay.joint_positions, "<f4", (frames, joints, 3), "joint_positions"
    )
    replay.plan_frames = _array(replay.plan_frames, "<u4", (plans,), "plan_frames")
    replay.plan_modes = _array(replay.plan_modes, "<i4", (plans,), "plan_modes")
    replay.plan_valid_lengths = _array(
        replay.plan_valid_lengths, "<u4", (plans,), "plan_valid_lengths"
    )
    replay.target_positions = _array(
        replay.target_positions, "<f4", (plans, TARGET_FRAMES, joints, 3),
        "target_positions",
    )
    if np.any(replay.plan_frames >= frames) or np.any(np.diff(replay.plan_frames) <= 0):
        raise ValueError("plan frames must be strictly increasing and inside the session")
    return replay


def write_replay(path: Path, replay: Replay) -> None:
    replay = validate(replay)
    header = HEADER.pack(
        MAGIC, VERSION, replay.fps, replay.frame_count, replay.joint_count,
        replay.qpos_count, replay.plan_count, TARGET_FRAMES, 0,
    )
    with path.open("xb") as stream:
        stream.write(header)
        for value in (
            replay.parents, replay.modes, replay.frame_plans, replay.qpos,
            replay.joint_positions, replay.plan_frames, replay.plan_modes,
            replay.plan_valid_lengths, replay.target_positions,
        ):
            stream.write(value.tobytes(order="C"))


def read_replay(path: Path) -> Replay:
    raw = path.read_bytes()
    if len(raw) < HEADER.size:
        raise ValueError("replay is shorter than its header")
    magic, version, fps, frames, joints, qpos, plans, target_frames, flags = HEADER.unpack_from(raw)
    if magic != MAGIC or version != VERSION or target_frames != TARGET_FRAMES or flags != 0:
        raise ValueError("unsupported replay header")
    if not (0 < frames <= 1_000_000 and 0 < joints <= 64 and 0 < qpos <= 256 and 0 < plans <= frames):
        raise ValueError("replay header dimensions are outside supported bounds")
    offset = HEADER.size

    def take(dtype, shape: tuple[int, ...], name: str) -> np.ndarray:
        nonlocal offset
        count = int(np.prod(shape))
        size = count * np.dtype(dtype).itemsize
        if size > len(raw) - offset:
            raise ValueError(f"replay is truncated in {name}")
        result = np.frombuffer(raw, dtype=dtype, count=count, offset=offset).reshape(shape).copy()
        offset += size
        return result

    replay = Replay(
        fps=fps,
        parents=take("<i4", (joints,), "parents"),
        modes=take("<i4", (frames,), "modes"),
        frame_plans=take("<i4", (frames,), "frame_plans"),
        qpos=take("<f4", (frames, qpos), "qpos"),
        joint_positions=take("<f4", (frames, joints, 3), "joint_positions"),
        plan_frames=take("<u4", (plans,), "plan_frames"),
        plan_modes=take("<i4", (plans,), "plan_modes"),
        plan_valid_lengths=take("<u4", (plans,), "plan_valid_lengths"),
        target_positions=take(
            "<f4", (plans, TARGET_FRAMES, joints, 3), "target_positions"
        ),
    )
    if offset != len(raw):
        raise ValueError(f"replay has {len(raw) - offset} trailing bytes")
    return validate(replay)

File: src/session_replay.cpp
#include "session_replay.hpp"

#include <array>
#include <bit>
#include <cmath>
#include <fstream>
#include <limits>
#include <stdexcept>
#include <string>
#include <type_traits>

namespace motionbricks::replay {
namespace {

constexpr std::array<unsigned char, 8> magic{'M', 'B', 'R', 'P', 'L', 'Y', '1', 0};
constexpr std::uint32_t version = 1;

class reader {
public:
    explicit reader(const std::filesystem::path & path) : stream_(path, std::ios::binary) {
        if (!stream_) throw std::runtime_error("could not open replay: " + path.string());
    }

    std::uint32_t u32(const char * field) {
        std::array<unsigned char, 4> bytes{};
        exact(bytes.data(), bytes.size(), field);
        return std::uint32_t(bytes[0]) | (std::uint32_t(bytes[1]) << 8U) |
               (std::uint32_t(bytes[2]) << 16U) | (std::uint32_t(bytes[3]) << 24U);
    }

    std::int32_t i32(const char * field) { return std::bit_cast<std::int32_t>(u32(field)); }
    float f32(const char * field) { return std::bit_cast<float>(u32(field)); }

    void exact(void * destination, std::size_t bytes, const char * field) {
        stream_.read(static_cast<char *>(destination), static_cast<std::streamsize>(bytes));
        if (stream_.gcount() != static_cast<std::streamsize>(bytes)) {
            throw std::runtime_error(std::string("replay is truncated in ") + field);
        }
    }

    void require_eof() {
        char trailing{};
        if (stream_.get(trailing)) throw std::runtime_error("replay has trailing bytes");
        if (!stream_.eof()) throw std::runtime_error("could not finish reading replay");
    }

private:
    std::ifstream stream_;
};

std::size_t checked_count(std::initializer_list<std::uint32_t> dimensions, const char * field) {
    std::size_t result = 1;
    for (const auto dimension : dimensions) {
        if (dimension != 0 && result > std::numeric_limits<std::size_t>::max() / dimension) {
            throw std::runtime_error(std::string("replay dimensions overflow in ") + field);
        }
        result *= dimension;
    }
    return result;
}

template <typename T, typename Read>
std::vector<T> read_values(reader & input, std::size_t count, Read read, const char * field) {
    std::vector<T> values;
    values.reserve(count);
    for (std::size_t index = 0; index < count; ++index) values.push_back((input.*read)(field));
    return values;
}

void finite(const std::vector<float> & values, const char * field) {
    for (const float value : values) {
        if (!std::isfinite(value)) throw std::runtime_error(std::string(field) + " contains a non-finite value");
    }
}

} // namespace

const float * session::frame_joints(std::uint32_t frame) const {
    if (frame >= frame_count) throw std::out_of_range("replay frame is out of bounds");
    return joint_positions.data() + checked_count({frame, joint_count, 3}, "frame offset");
}

const float * session::plan_targets(std::uint32_t plan) const {
    if (plan >= plan_count) throw std::out_of_range("replay plan is out of bounds");
    return target_positions.data() + checked_count({plan, target_frames, joint_count, 3}, "plan offset");
}

session load(const std::filesystem::path & path) {
    reader input(path);
    std::array<unsigned char, 8> actual_magic{};
    input.exact(actual_magic.data(), actual_magic.size(), "magic");
    if (actual_magic != magic) throw std::runtime_error("unsupported replay magic");
    if (input.u32("version") != version) throw std::runtime_error("unsupported replay version");

    session result;
    result.fps = input.u32("fps");
    result.frame_count = input.u32("frame count");
    result.joint_count = input.u32("joint count");
    result.qpos_count = input.u32("qpos count");
    result.plan_count = input.u32("plan count");
    const auto targets = input.u32("target frame count");
    const auto flags = input.u32("flags");
    if (result.fps == 0 || result.fps > 1000 || result.frame_count == 0 ||
        result.frame_count > 1'000'000 || result.joint_count == 0 || result.joint_count > 64 ||
        result.qpos_count == 0 || result.qpos_count > 256 || result.plan_count == 0 ||
        result.plan_count > result.frame_count) {
        throw std::runtime_error("replay header dimensions are outside supported bounds");
    }
    if (targets != target_frames || flags != 0) throw std::runtime_error("unsupported replay header");

    const auto payload_words = checked_count({result.joint_count}, "payload") +
        checked_count({result.frame_count, 2}, "payload") +
        checked_count({result.frame_count, result.qpos_count}, "payload") +
        checked_count({result.frame_count, result.joint_count, 3}, "payload") +
        checked_count({result.plan_count, 3}, "payload") +
        checked_count({result.plan_count, target_frames, result.joint_count, 3}, "payload");
    constexpr std::size_t max_payload_words = (512ULL * 1024ULL * 1024ULL - 40ULL) / 4ULL;
    if (payload_words > max_payload_words) throw std::runtime_error("replay payload exceeds the size limit");

    result.parents = read_values<std::int32_t>(input, result.joint_count, &reader::i32, "parents");
    result.modes = read_values<std::int32_t>(input, result.frame_count, &reader::i32, "modes");
    result.frame_plans = read_values<std::int32_t>(input, result.frame_count, &reader::i32, "frame plans");
    result.qpos = read_values<float>(input, checked_count({result.frame_count, result.qpos_count}, "qpos"), &reader::f32, "qpos");
    result.joint_positions = read_values<float>(
        input, checked_count({result.frame_count, result.joint_count, 3}, "joint positions"),
        &reader::f32, "joint positions");
    result.plan_frames = read_values<std::uint32_t>(input, result.plan_count, &reader::u32, "plan frames");
    result.plan_modes = read_values<std::int32_t>(input, result.plan_count, &reader::i32, "plan modes");
    result.plan_valid_lengths = read_values<std::uint32_t>(input, result.plan_count, &reader::u32, "plan lengths");
    result.target_positions = read_values<float>(
        input, checked_count({result.plan_count, target_frames, result.joint_count, 3}, "target positions"),
        &reader::f32, "target positions");
    input.require_eof();

    if (result.parents.front() != -1) throw std::runtime_error("replay root parent must be -1");
    for (std::uint32_t joint = 1; joint < result.joint_count; ++joint) {
        if (result.parents[joint] < 0 || result.parents[joint] >= static_cast<std::int32_t>(joint)) {
            throw std::runtime_error("replay joint topology is invalid");
        }
    }
    for (const auto plan : result.frame_plans) {
        if (plan < -1 || plan >= static_cast<std::int32_t>(result.plan_count)) {
            throw std::runtime_error("replay frame plan is out of bounds");
        }
    }
    for (std::uint32_t plan = 0; plan < result.plan_count; ++plan) {
        if (result.plan_frames[plan] >= result.frame_count ||
            (plan > 0 && result.plan_frames[plan] <= result.plan_frames[plan - 1])) {
            throw std::runtime_error("replay plan frames are invalid");
        }
    }
    finite(result.qpos, "qpos");
    finite(result.joint_positions, "joint positions");
    finite(result.target_positions, "target positions");
    return result;
}

} // namespace motionbricks::replay
